agent/stats/container.go

```diff
@@ -0,0 +1,128 @@
+	SleepBetweenUsageDataCollection = 500 * time.Millisecond
+
+	// ContainerStatsBufferLength is the number of usage metrics stored in memory for a container. It is calculated as
+	// Number of usage metrics gathered in a second (10) * 60 * Time duration in minutes to store the data for (2)
+	ContainerStatsBufferLength = 1200
+)
+
+// ContainerStatsCollector defines methods to get container stats. This interface is defined to
+// make testing easier.
+type ContainerStatsCollector interface {
+	getContainerStats(container *CronContainer) (*ContainerStats, error)
+}
+
+// LibcontainerStatsCollector implements ContainerStatsCollector.
+type LibcontainerStatsCollector struct{}
+
+// dockerGraphPath stores the docker exec driver path.
+var dockerGraphPath string
+
+func init() {
+	dockerGraphPath = utils.DefaultIfBlank(os.Getenv(DockerGraphPath), DefaultDockerGraphPath)
+}
+
+// StartStatsCron starts a go routine to periodically pull usage data for the container.
+func (container *CronContainer) StartStatsCron() {
+	// Create the queue to store utilization data from cgroup fs.
+	container.statsQueue = NewQueue(ContainerStatsBufferLength)
+
+	// Create the context to handle deletion of container from the manager.
+	// The manager can cancel the cronStats go routing by calling StopStatsCron method.
+	container.ctx, container.cancel = context.WithCancel(context.Background())
+	go container.cronStats()
+}
+
+// StopStatsCron stops the periodic collection of usage data for the container..
+func (container *CronContainer) StopStatsCron() {
+	container.cancel()
+}
+
+// newCronContainer creates a CronContainer object.
+func newCronContainer(dockerID *string, name *string) *CronContainer {
+	statePath := filepath.Join(dockerGraphPath, DockerExecDriverPath, *dockerID)
+
+	container := &CronContainer{
+		containerMetadata: &ContainerMetadata{
+			DockerID: dockerID,
+			Name:     name,
+		},
+		statePath: statePath,
+	}
+
+	container.statsCollector = &LibcontainerStatsCollector{}
+	return container
+}
+
+// cronStats periodically pulls usage data for the container from cgroup fs.
+func (container *CronContainer) cronStats() {
+	for {
+		select {
+		case <-container.ctx.Done():
+			return
+		default:
+			stats, err := container.statsCollector.getContainerStats(container)
+			if err != nil {
+				log.Debug("Error getting stats", "error", err, "contianer", container)
+			} else {
+				container.statsQueue.Add(stats)
+			}
+			time.Sleep(SleepBetweenUsageDataCollection)
+		}
+	}
+}
+
+// getContainerStats reads usage data of a container from the cgroup fs.
+func (collector *LibcontainerStatsCollector) getContainerStats(container *CronContainer) (*ContainerStats, error) {
+	state, err := libcontainer.GetState(container.statePath)
+	if err != nil {
+		// The state file is not created immediately when a container starts.
+		// Bubble up the error.
+		return nil, err
+	}
+	containerStats, err := libcontainer.GetStats(nil, state)
```

Comment:
The first argument to licontainer.GetStats() is never used in that method. I'll add a comment.